yarn-project/prover-client/src/orchestrator/orchestrator.ts

```diff
@@ -0,0 +1,516 @@
+    logger.info(`Starting new block with ${numTxs} transactions`);
+    // we start the block by enqueueing all of the base parity circuits
+    let baseParityInputs: BaseParityInputs[] = [];
+    let l1ToL2MessagesPadded: Tuple<Fr, typeof NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP>;
+    try {
+      l1ToL2MessagesPadded = padArrayEnd(l1ToL2Messages, Fr.ZERO, NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP);
+    } catch (err) {
+      throw new Error('Too many L1 to L2 messages');
+    }
+    baseParityInputs = Array.from({ length: NUM_BASE_PARITY_PER_ROOT_PARITY }, (_, i) =>
+      BaseParityInputs.fromSlice(l1ToL2MessagesPadded, i),
+    );
+
+    //TODO:(@PhilWindle) Temporary until we figure out when to perform L1 to L2 insertions to make state consistency easier.
+    await Promise.resolve();
+
+    const promise = new Promise<ProvingResult>((resolve, reject) => {
+      this.provingState = new ProvingState(
+        numTxs,
+        resolve,
+        reject,
+        globalVariables,
+        l1ToL2Messages,
+        baseParityInputs.length,
+        emptyTx,
+      );
+    }).catch((reason: string) => ({ status: PROVING_STATUS.FAILURE, reason } as const));
+
+    for (let i = 0; i < baseParityInputs.length; i++) {
+      this.enqueueJob(this.provingState!.Id, PROVING_JOB_TYPE.BASE_PARITY, () =>
+        this.runBaseParityCircuit(baseParityInputs[i], i, this.provingState!.Id),
+      );
+    }
+
+    const ticket: ProvingTicket = {
+      provingPromise: promise,
+    };
+    return ticket;
+  }
+
+  /**
+   * The interface to add a simulated transaction to the scheduler
+   * @param tx - The transaction to be proven
+   */
+  public async addNewTx(tx: ProcessedTx): Promise<void> {
+    if (!this.provingState) {
+      throw new Error(`Invalid proving state, call startNewBlock before adding transactions`);
+    }
+
+    if (this.provingState.numRealTxs === this.provingState.transactionsReceived) {
+      throw new Error(`Rollup already contains ${this.provingState.transactionsReceived} transactions`);
+    }
+
+    validateTx(tx);
+
+    logger.info(`Received transaction :${tx.hash}`);
+
+    // We start the transaction by enqueueing the state updates
+
+    const txIndex = this.provingState!.addNewTx(tx);
+    // we start this transaction off by performing it's tree insertions and
+    await this.prepareBaseRollupInputs(BigInt(txIndex), tx, this.provingState!.globalVariables, this.provingState!.Id);
+
+    if (this.provingState.transactionsReceived === this.provingState.numRealTxs) {
+      // we need to pad the rollup with empty transactions
+      const numPaddingTxs = this.provingState.numPaddingTxs;
+      for (let i = 0; i < numPaddingTxs; i++) {
+        const paddingTxIndex = this.provingState.addNewTx(this.provingState.emptyTx);
+        await this.prepareBaseRollupInputs(
+          BigInt(paddingTxIndex),
+          this.provingState!.emptyTx,
+          this.provingState!.globalVariables,
+          this.provingState!.Id,
+        );
+      }
```

Comment:
I'm really unsure about the interface between orchestrator and provingstate. 

It feels to me as though they are 1-1: we don't really want people calling `startNewBlock` while there is unfinished state, 
plus this class makes use of intimate knowledge of the internals of provingstate (e.g. `storeMergeInputs` below).